src/Adapter/ContextStateManager.php

```diff
@@ -0,0 +1,192 @@
+<?php
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter;
+
+use Cart;
+use Context;
+use Country;
+use Currency;
+use Customer;
+use Language;
+
+/**
+ * Allows manipulating context state.
+ * This was adapted for specific broken legacy use cases when the previous state of context must be restored after some actions.
+ *
+ * e.g. order creation from back office.
+ *  Legacy requires Context properties (currency, country etc.) instead of using cart properties
+ *  so some context props must be changed for a while and then restored to previous state.
+ */
+final class ContextStateManager
+{
+    /**
+     * @var Context
+     */
+    private $context;
+
+    /**
+     * @var Cart|null
+     */
+    private $cart;
```

Comment:
Rename all these fields into `savedCart` `savedCountry` ... So that we understand it is saved from the initial context